src/Motor.Extensions.Hosting.Kafka/KafkaMessageConsumer.cs

```diff
@@ -162,47 +167,62 @@ private void WriteStatistics(string json)
         }
     }
 
-    private void SingleMessageHandling(CancellationToken token, ConsumeResult<string?, byte[]> msg)
+    private async Task SingleMessageHandlingAsync(ConsumeResult<string?, byte[]> msg, CancellationToken token)
     {
-        _logger.LogDebug(
-            $"Received message from topic '{msg.Topic}:{msg.Partition}' with offset: '{msg.Offset}[{msg.TopicPartitionOffset}]'");
-        var cloudEvent = KafkaMessageToCloudEvent(msg.Message);
-
-        var taskAwaiter = ConsumeCallbackAsync?.Invoke(cloudEvent, token).GetAwaiter();
-        taskAwaiter?.OnCompleted(() =>
+        try
         {
-            var processedMessageStatus = taskAwaiter?.GetResult();
-            _messageSemaphore.Release();
-            switch (processedMessageStatus)
-            {
-                case ProcessedMessageStatus.Success:
-                    break;
-                case ProcessedMessageStatus.TemporaryFailure:
-                    break;
-                case ProcessedMessageStatus.InvalidInput:
-                    break;
-                case ProcessedMessageStatus.CriticalFailure:
-                    break;
-                case ProcessedMessageStatus.Failure:
-                    break;
-                default:
-                    throw new ArgumentOutOfRangeException();
-            }
-
-            if (msg.Offset % _options.CommitPeriod != 0)
-            {
-                return;
-            }
+            _logger.LogDebug(
+                "Received message from topic '{Topic}:{Partition}' with offset: '{Offset}[{TopicPartitionOffset}]'",
+                msg.Topic, msg.Partition, msg.Offset, msg.TopicPartitionOffset);
+            var cloudEvent = KafkaMessageToCloudEvent(msg.Message);
+
+            var retryPolicy = Policy
+                .HandleResult<ProcessedMessageStatus>(status => status == ProcessedMessageStatus.TemporaryFailure)
+                .WaitAndRetryAsync(_options.RetriesOnTemporaryFailure,
+                    retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt)));
+            var status = await retryPolicy.ExecuteAsync(() => ConsumeCallbackAsync!.Invoke(cloudEvent, token));
+            HandleMessageStatus(msg, status);
+        }
+        catch (Exception e)
+        {
+            _logger.LogCritical(e, "Unexpected exception in message handling");
```

Comment:
Not that important, but how about using a `LogEvent` here? (same for the `LogDebug` above in line 174).